core/kazoo_ast/src/kz_edoc_layout.erl

```diff
@@ -0,0 +1,1240 @@
+        _ -> String
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc Why EDoc is so stupid to not make a paragraph of last text/markups
+%% if they are not a block level elements?
+%% This is a copy pasta code from `edoc_wiki' with a change to remove
+%% stupid whitespace only `#xmlText{}'.
+%% @end
+%%------------------------------------------------------------------------------
+-spec normalize_description([#xmlElement{}], map()) -> [exporty_thing()].
+normalize_description([], _) ->
+    [];
+normalize_description(Es, Context) ->
+    normalize_links(make_paraghraph(Es, [], []), Context).
+
+-spec make_paraghraph([#xmlElement{}], [exporty_thing()], [exporty_thing()]) -> [exporty_thing()].
+make_paraghraph([E=#xmlElement{name = Name}|Es], As, Bs) ->
+    case is_block_element(Name) of
+        true -> par_flush(Es, As, [E | Bs]);
+        false -> make_paraghraph(Es, [E | As], Bs)
+    end;
+make_paraghraph([E=#xmlText{value = Value}|Es], As, Bs) ->
+    case is_only_whitespace(Value) of
+        true -> make_paraghraph(Es, As, Bs);
+        false ->
+            Normalized = E#xmlText{value = [C || C <- Value, C =/= $\n, C =/= $\r]},
+            make_paraghraph(Es, [Normalized | As], Bs)
+    end;
+make_paraghraph([], [], Bs) ->
+    lists:reverse(Bs);
+make_paraghraph([], As, Bs) ->
+    lists:reverse([{p, lists:reverse(As)} | Bs]).
+
+is_block_element('article') -> true;
+is_block_element('aside') -> true;
+is_block_element('canvas') -> true;
+is_block_element('dd') -> true;
+is_block_element('div') -> true;
+is_block_element('dl') -> true;
+is_block_element('dt') -> true;
+is_block_element('fieldset') -> true;
+is_block_element('figcaption') -> true;
+is_block_element('figure') -> true;
+is_block_element('footer') -> true;
+is_block_element('form') -> true;
+is_block_element('h1') -> true;
+is_block_element('h2') -> true;
+is_block_element('h3') -> true;
+is_block_element('h4') -> true;
+is_block_element('h5') -> true;
+is_block_element('h6') -> true;
+is_block_element('header') -> true;
+is_block_element('hgroup') -> true;
+is_block_element('hr') -> true;
+is_block_element('li') -> true;
+is_block_element('main') -> true;
+is_block_element('nav') -> true;
+is_block_element('noscript') -> true;
+is_block_element('ol') -> true;
+is_block_element('output') -> true;
+is_block_element('p') -> true;
+is_block_element('pre') -> true;
+is_block_element('section') -> true;
+is_block_element('table') -> true;
+is_block_element('tfoot') -> true;
+is_block_element('ul') -> true;
+is_block_element('video') -> true;
+is_block_element(_) -> false.
+
+-spec par_flush([#xmlElement{}], [exporty_thing()], [exporty_thing()]) -> [exporty_thing()].
+par_flush(Es, [], Bs) ->
+    make_paraghraph(Es, [], Bs);
+par_flush(Es, As, Bs) ->
+    make_paraghraph(Es, [], Bs ++ [{p, As}]).
+
+-spec get_elem(atom(), [#xmlElement{}]) -> [#xmlElement{}].
+get_elem(Name, [#xmlElement{name = Name} = E | Es]) ->
```

Comment:
Can you look at using/expanding kz_xml for these types of operations? Would be good to expand that functionality for others to build on as well.

Reply:
I can, but don't forget that this file is exactly `edoc_layout` module from `edoc` app itself with some changes like return proplist instead of html and specs added and some other minor changes.

And currently it doesn't have any dependency on anything other than erlang itself.

Reply:
Is there a reason to avoid Kazoo modules as dependencies though?